test/test_functionalization.py

```diff
@@ -85,6 +85,25 @@ def f(x):
         [1., 1.],
         [1., 1.]]))""")
 
+    def test_tensor_list_composite(self):
+        def f(x):
+            # Test an op with TensorList input
+            y = torch.block_diag(x, x)
+            return y
+        self.assert_functionalization(f, torch.ones(2, 2))
+        logs = self.get_logs(f, torch.ones(2, 2))
+        # Only seeing copy_() calls in the logs are actually expected:
+        # - block_diag is a CompositeImplicitAutograd op, implemented in terms of copy_() and a few other ops.
+        # - copy_() doesn't have an out-of-place variant, so the pass leaves it alone
```

Comment:
Wait, so when do we turn copy_() into .to(...).expand_as(...) ? Is this a bug?

Reply:
This is actually expected, but I just realized that it will pose some problems for LazyTensor/XLA/Vulkan later.

This happens because `block_diag` doesn't decompose during the functionalization pass: instead, `block_diag` will hit the functionalization boxed fallback, redispatch, and decompose **underneath** the pass (so its base ops will not be functionalized).

The easiest solution is to just say that all `CompositeImplicitAutograd` ops should decompose during functionalization. I remember trying that initial and running into some problems - I forget exactly what they were, but one I can think of is that this would prevent XLA/LTC/Vulkan from providing their own custom implementations of composite ops.

We could actually maybe get around that in a way similar to how we deal with the `Autograd` key right now - by making functionalization per backend concept, registering everything to an alias `Functionalization` key, and letting backends override composite kernels by registering their own implementations to per-backend keys.

Another simpler option would just be to do the same thing that we're doing for functorch vmap in core: maintain a list of "problematic" composite that get explicit functionalization kernels.

I think this probably doesn't need to be addressed for now, but will definitely come up next year when trying to integrate the pass with LazyTensor.

But the downside is that backends can't "recover" the composite op after it's decomposed, if they want to provide a custom kernel. I think we could actually get around this similar to how

Reply:
> This happens because block_diag doesn't decompose during the functionalization pass: instead, block_diag will hit the functionalization boxed fallback, redispatch, and decompose underneath the pass (so its base ops will not be functionalized).

Oh that makes a lot of sense, thank you for the clarification.

> But the downside is that backends can't "recover" the composite op after it's decomposed, if they want to provide a custom kernel. I think we could actually get around this similar to how

This is sounding very similar to the vmap(vjp issue we were discussing earlier today

Reply:
It definitely does 😬 . "An op has (potentially multiple) decompositions, that should be toggleable/customizable".

Alias keys are one tool that we have to solve this type of problem: We register the "vanilla" decomposition to the alias key, and allow backends to customize it by registering something to their own per-backend key instead. That doesn't work for functorch though, since functorch isn't a "backend". Maybe there's something there though.